src/main/java/io/confluent/idesidecar/websocket/resources/WebsocketEndpoint.java

```diff
@@ -0,0 +1,312 @@
+      log.error("Invalid websocket message header originator value -- not an integer: " + headers.originator + ". Removing and closing session.");
+      sessions.remove(senderSession);
+      senderSession.close();
+      return;
+    }
+
+    if (claimedWorkspaceId != workspaceSession.processId()) {
+      log.error("Workspace " + workspaceSession.processId() + " sent message with incorrect originator value: " + claimedWorkspaceId + ".Removing and closing session.");
+      sessions.remove(senderSession);
+      senderSession.close();
+      return;
+    }
+
+    log.debug("Received " + headers.type + " message from workspace: " + workspaceSession.processId());
+
+    // Handle the message based on its audience.
+    if (headers.audience == Audience.workspaces)
+    {
+      // Message is intended for all (other) workspaces, using sidecar as a broadcast bus
+      // for workspace-to-all-other-workspace messages.
+      int otherCount = sessions.size() - 1;
+      if (otherCount == 0) {
+        log.debug("No other workspaces to broadcast message to.");
+        return;
+      } else {
+        log.debug("Broadcasting message to " + otherCount + " other workspaces");
+        broadcast(m, workspaceSession);
+      }
+
+    } else if (headers.audience == Audience.sidecar) {
+      // Message must be intended for sidecar
+
+      // todo defer to an internal message router here.
+
+      log.error("Unexpected message audience: " + m.getHeaders().audience);
+    } else {
+      // We don't (yet) support or design for directed workspace -> workspace messages.
+      log.error("Unhandled message audience: " + m.getHeaders().audience);
+    }
+  }
+
+  @OnOpen
+  public void onOpen(Session session) throws IOException {
+    // Don't store into sessions map until successful handling of ACCESS_REQUEST message,
+    // so do nothing of importance here.
+    log.info("New websocket session opened: " + session.getId());
+    // Request must have had a valid access token to pass through AccessTokenFilter, so we can assume that the session is authorized.
+    // The workspace process id should have been passed as a request parameter, though.
+    String workspaceIdString = session.getRequestParameterMap().get("workspace_id").get(0);
+
+    if (workspaceIdString == null) {
+      log.error("No workspace_id parameter provided. Closing session.");
+      session.close();
+      return;
+    }
+
+    int workspaceId;
+    try {
+      workspaceId = Integer.parseInt(workspaceIdString);
+    } catch (NumberFormatException e) {
+      log.error("Invalid workspace_id parameter value: " + workspaceIdString + ". Closing session.");
+      session.close();
+      return;
+    }
+
+    log.info("New websocket session opened for workspace pid: " + workspaceId);
+    // create new WorkspaceSession object and store in sessions map.
+    WorkspaceSession newWorkspaceSession = new WorkspaceSession(workspaceId);
+    sessions.put(session, newWorkspaceSession);
```

Comment:
Do we need to check if this is a known workspace ID that's managed by the `KnownWorkspacesBean`?

Reply:
Yes, thanks, was planning on that. Done, and clicktested:

2024-12-13 12:07:23,903 ERROR [io.con.ide.web.res.WebsocketEndpoint] (vert.x-eventloop-thread-2) Unauthorized workspace id: 1122. Closing session.

Reply:
In the future, I think we can completely invert the handshaking process, etc. so that the authoritative establishment process is the websocket connection, simplifying things elesewhere. But this is self-consistent at least now. Can migrate / invert things in the future.

And then as long as vscode/extension-side has a happy websocket, then getting at a SidecarHandle can become round-trip free (@shouples).